File: src/leetcode/leetcode49/Solution.java
package leetcode.leetcode49;

import java.util.*;

public class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();
        for(String str : strs){
            char[] arr = str.toCharArray();
            Arrays.sort(arr);
            String key = new String(arr);

            if( map.containsKey(key) ){
                map.get(key).add(str);
            }else {
                map.put(key,new ArrayList<>());
                map.get(key).add(str);
            }
        }
        return  new ArrayList<>(map.values());
    }

    public static void main(String[] args) {
        String[] strs = new String[]{"eat","tea","tan","ate","nat","bat"};
        new Solution().groupAnagrams(strs);
    }
}

File: src/leetcode/leetcode49/Solution.py
from collections import defaultdict
from typing import List


"""
学习使用defaultdict，sorted，d.append list()
"""
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        d = defaultdict(list)
        for s in strs:
            d[''.join(sorted(s))].append(s)
        return list(d.values())

if __name__ == '__main__':
    strs = ["eat","tea","tan","ate","nat","bat"]
    list = Solution().groupAnagrams(strs)
    print(list)